### ISSUE TITLE
Training stalls during evaluation

### ISSUE THREAD

**Issue by TimurKhayrullin** [facebookresearch/demucs#462]
## ❓ Questions

Hi There, I'm currently trying to train htdemucs in Google Colab, and I have successfully started training with dora. However, the script is getting stuck after 48/50 evaluation steps. It just stalls and eats compute units. Could someone help me figure out the issue? Any advice is appreciated. Here is my config.yaml: 

 
`# @package _global_

model: htdemucs

dset:
  full_cv: true
  segment: 11

epochs: 1
batch_size: 4
max_batches: 1

misc:
  num_workers : 0
  verbose: true

test:
  workers: 1`

Thanks!

**Comment by gold90123** [comment]
Does it said Killed?

**Comment by adefossez** [comment]
The SDR metric is really costly to compute, especially in a Colab that doesn’t have multiple CPUs you can remove it with test.sdr=false. You can also change test.every so that eval doesn’t happen so often (set it to the number of epochs). It will still happen at the very end but the model will be mostly trained.

**Comment by TimurKhayrullin** [comment]
> Does it said Killed?

No, It just stalled



> The SDR metric is really costly to compute, especially in a Colab that doesn’t have multiple CPUs you can remove it with test.sdr=false. You can also change test.every so that eval doesn’t happen so often (set it to the number of epochs). It will still happen at the very end but the model will be mostly trained.

Tried setting test.sdr to false, it still stalled. What ended up working was setting segment to 2, but the results weren't that great. At the moment we are working with 1 GPU with 40 GB of vram, what would be your recommended experiment config to get the best training results given our hardware?

**Comment by adefossez** [comment]
that is a bit surprising, 40 GB of VRAM should be enough to train with a batch size of 4 and segment length of 10 seconds.
You can look into `demucs/grids/mmi.py`, for some hyperparams, e.g. maybe
```
test:
    shifts: 0
htdemucs:
    dconv_mode: 3
    depth: 4
    t_layers: 5
    t_dropout: 0.02
    bottom_channels: 512
```

We trained the model we released on 32GB GPUs, (but used 8 of them hence the original batch size of 32).

**Comment by njusq** [comment]
I also got stuck in test stage. I printed the length of reference signals, and found the wrong signals came from utils/evaluate.py:
`
test_set = musdb.DB(args['dset']['musdb'], subsets=["test"], is_wav=True)
`
which uses _read_ffmpeg to get signal. The length of variable 'buffer' is shorter than which it should be.
![image](https://github.com/facebookresearch/demucs/assets/38641314/6d9152bc-0291-4aea-aeb9-5832342a0f66)

**Comment by EnriqueSampaio** [comment]
@njusq did you found any solution for this? My train got stuck during the evaluation too, it was not the last epoch, so the old SDR shouldn't be the problem.

The curious thing is that when I evaluate the model separately (tools.test_pretrained) it uses the same method but runs without problems.

**Comment by njusq** [comment]
> 

Not yet.
Test failure not always happened to me but it indeed affected my experiment efficiency.
Several tips you may refer to:
1. Turn to a machine with larger memory.
2. Modify conf['test']['every'] to a larger number (e.g. 100) in the configuration file "conf/config.yaml", it reduces the test times during training.

**Comment by njusq** [comment]
@EnriqueSampaio 
Hi～
It was stuck in loading target tracks. I solved the problem by retrying many times. 
My modified code in utils/evaluate.py:
```
retry_times = 10
for _ in range(retry_times):
  try:
    references = th.stack(
      [th.from_numpy(track.targets[name].audio).t() for name in model.sources])
    break
  except Exception as e:
    logger.error(f"Error occurred: {e}")
    if _ < retry_times - 1:
      logger.info(f"Retry...")
    else:
      logger.info("Maximum retry times reached. Exiting...")
      break
```
